include/net/xdp.h

```diff
@@ -63,17 +63,26 @@ struct xdp_rxq_info {
 	struct xdp_mem_info mem;
 } ____cacheline_aligned; /* perf critical, avoid false-sharing */
 
+#ifdef CONFIG_XDP_LUA
+struct xdplua {
+	struct lua_State	*L;
+	spinlock_t		*lock;
```

Comment:
Is `spinlock` the right concurrency primitive here?
So this is a lock per lua state, and there is a lua state per CPU, and spinlocks are only appropriate for locking accesses between different CPUs to the same data. And also, even in that context, spinlocks are highly inefficient when they are taken for a long time, such as would be the case for executing a lua script.

So if you ever happened upon a case where the same CPU got this lock multiple times, it would have hung forever, so I suspect that you might not even need a lock at all.

Reply:
Unfortunately a lock is indeed needed. If a user attempts to load a Lua script while another script is being executed in the same Lua state, we will have concurrent accesses to the a Lua State, which will cause synchronization issues. If not a `spinlock`, what would you suggest?

Reply:
Based on your other message, I see now that what you probably need is just `local_bh_disable` and `local_bh_enable` when loading script.